File: 2022/elixir_2022/lib/day_02.ex
file =
  File.read!("../inputs/day_02.txt")
  |> String.split("\n")

myMoveMap = %{
  "X" => 0,
  "Y" => 1,
  "Z" => 2
}

opponentMoveMap = %{
  "A" => 0,
  "B" => 1,
  "C" => 2
}

defmodule Day02 do
  def computeScore(opponentMove, myMove) do
    winningMove = (opponentMove + 1) |> rem(3)
    score = myMove + 1

    cond do
      myMove == winningMove -> score + 6
      myMove == opponentMove -> score + 3
      true -> score
    end
  end
end

# Part 1
file
|> Enum.map(fn line ->
  [opponentMove, myMove] = String.split(line)
  Day02.computeScore(opponentMoveMap[opponentMove], myMoveMap[myMove])
end)
|> Enum.sum()
|> IO.inspect()

# Part 2
file
|> Enum.map(fn line ->
  [opponentMove, gameResult] = String.split(line)

  myMove =
    cond do
      gameResult == "X" -> (opponentMoveMap[opponentMove] + 2) |> rem(3)
      gameResult == "Y" -> opponentMoveMap[opponentMove]
      gameResult == "Z" -> (opponentMoveMap[opponentMove] + 1) |> rem(3)
    end

  Day02.computeScore(opponentMoveMap[opponentMove], myMove)
end)
|> Enum.sum()
|> IO.inspect()

File: 2022/python_2022/day_02.py
file = open("../inputs/day_02.txt", "r")

my_move_map = {
    "X": 0,
    0: "X",
    "Y": 1,
    1: "Y",
    "Z": 2,
    2: "Z",
}

opponent_move_map = {
    "A": 0,
    0: "A",
    "B": 1,
    1: "B",
    "C": 2,
    2: "C",
}


def compute_score(opponent_move, my_move) -> int:
    score = my_move_map[my_move] + 1

    winning_move = (opponent_move_map[opponent_move] + 1) % 3
    if my_move_map[my_move] == opponent_move_map[opponent_move]:
        return score + 3
    elif my_move_map[my_move] == winning_move:
        return score + 6
    else:
        return score


p1_score, p2_score = 0, 0
for line in file:
    opponent_move, my_move = line.strip().split(" ", 2)
    p1_score += compute_score(opponent_move, my_move)

    game_result = my_move
    if game_result == "X":
        my_move = (opponent_move_map[opponent_move] + 2) % 3
    elif game_result == "Y":
        my_move = opponent_move_map[opponent_move]
    else:
        my_move = (opponent_move_map[opponent_move] + 1) % 3

    p2_score += compute_score(opponent_move, my_move_map[my_move])

print(f"Part 1: {p1_score}")
print(f"Part 2: {p2_score}")
